src/Npgsql/NpgsqlConnector.cs

```diff
@@ -563,6 +596,36 @@ internal async Task Open(NpgsqlTimeout timeout, bool async, CancellationToken ca
             TypeMapper.Bind(DatabaseInfo);
         }
 
+
+        internal void UpdateServerPrimaryStatus()
+        {
+            StartUserAction();
+            WritePregenerated(PregeneratedMessages.ServerIsSecondary);
+            Flush();
+
+            var columnsMsg = ReadMessage();
+            var rowMsg = Expect<DataRowMessage>(ReadMessage(), this);
+
+            var columnCount = ReadBuffer.ReadInt16();
+            var lengthOfBooleanColumn = ReadBuffer.ReadInt32();
+            var resultSetBuffer = new byte[lengthOfBooleanColumn];
+            ReadBuffer.ReadBytes(resultSetBuffer, 0, lengthOfBooleanColumn);
```

Comment:
This look pretty hardcore - we probably should rewrite this to a higher-level API (do not yet have any ideas on how it should look).

Reply:
Yeah, absolutely, I wasn't able to find a way of running an actual query and getting a result set with the objects I had in the state they're in working; any suggestions as to how to actually make this not filth, much appreciated!

Reply:
https://github.com/npgsql/npgsql/pull/2771#issuecomment-569947899